src/cargo/util/toml/mod.rs

```diff
@@ -97,11 +109,69 @@ fn read_manifest_from_str(
 
     let mut unused = BTreeSet::new();
     let deserializer = toml::de::Deserializer::new(contents);
-    let manifest: manifest::TomlManifest = serde_ignored::deserialize(deserializer, |path| {
+    let manifest: manifest::TomlManifest = match serde_ignored::deserialize(deserializer, |path| {
         let mut key = String::new();
         stringify(&mut key, &path);
         unused.insert(key);
-    })?;
+    }) {
+        Ok(manifest) => manifest,
+        Err(e) => {
+            return if let Some(span) = e.span() {
+                let (line, column) = translate_position(contents.as_bytes(), span.start);
+                let lines = contents.lines();
+                let line_count = lines.clone().count();
+                let mut source = lines.skip(line).next().unwrap_or_default().to_string();
+                // Add back the new line that was stripped by `.lines()`
+                if line_count > line + 1 {
+                    source.push('\n');
+                // If we are trying to highlight past the end of the file, add a
+                // placeholder character to the end of the line.
+                } else if span.end >= source.len() - 1 {
+                    source.push('∅');
+                }
```

Comment:
1. This should be in annotate-snippets,  not cargo
2. The `'\n'` should be `"\\n"` so it gets rendered
3. These should be Dimmed to indicate they aren't literal text

Reply:
I'm also not clear why the `∅` is being added, can you say more about the motivation there? I'm a little concerned about any potential confusion if users are not familiar with what that symbol means. I don't think rustc typically displays errors like that, and I think it is reasonably clear without a special character. For example:

```
error: this file contains an unclosed delimiter
 --> src/main.rs:4:10
  |
4 | fn x() {
  |        - ^
  |        |
  |        unclosed delimiter
```

or

```
error: expected one of `->`, `where`, or `{`, found `<eof>`
 --> src/main.rs:4:9
  |
4 | fn main()
  |         ^ expected one of `->`, `where`, or `{`
```

Reply:
I'm trying to find the conversation around invisible characters but it is a bit buried.

I *think* an aspect is that TOML is newline sensitive, unlike Rust, which elevates the importance of invisible characters and makes the chance of an error more frequent.

Reply:
I removed `cargo`'s ability to render invisible characters, as it is something that should be discussed in another issue